Add keyboard shortcuts to prompt

zevisert opened the pull request:
<!--
Thanks for sending a pull request!
Please fill in the following content to let us know better about this change.
-->

## Description
<!-- Describe what the change is -->

See the suggestion in #380.

- feat(prompt): add keyboard shortcuts with config option
  > Also added chore option to conventional_commits choices

This PR adds a config option called `use_shortcuts` which enables shortcuts in conventional commit choices. It defaults to false, so it'd be opt-in for now. In #380, the suggestion was to add the `use_shortcuts: bool` kwarg to questionary.prompt in [commands/commit.py](https://github.com/zevisert/commitizen/blob/65eb4f8b53f2e050b5ddbafa4b32439ef2f1a13d/commitizen/commands/commit.py#L50), but questionary passes this on to every question, and subsequently throws for prompt types that don't support `use_shortcuts`. The solution was to add `use_shortcuts` to the first question in [cz/conventional_commits/conventional_commits.py](https://github.com/zevisert/commitizen/blob/65eb4f8b53f2e050b5ddbafa4b32439ef2f1a13d/commitizen/cz/conventional_commits/conventional_commits.py#L48) instead.

## Checklist

- [ ] Add test cases to all the changes you introduce
- [ ] Run `./script/format` and `./script/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [ ] Update the documentation for the changes

zevisert commented:
I'll add tests later if you like the change. I noticed there wasn't complete approval in #380 yet.

Lee-W commented on commitizen/cz/conventional_commits/conventional_commits.py:
```diff
@@ -45,54 +45,76 @@ def questions(self) -> List[Dict[str, Any]]:
+                        "key": "c",
+                    },
+                    {
+                        "value": "chore",
```
We decide not to add "chore" as this is not required by conventional commit https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type

Lee-W commented:
@zevisert Thanks for you contribution! This feature looks good to me @Woile What do you think?

Woile commented:
This is great! I like it!

zevisert commented:
Great, I'll follow up with some tests and docs in the next day or two - and sure I'll drop that chore option. I didn't mean to submit that here, that's just from me fiddling around.

zevisert commented:
Hey sorry for the delay! I rebased out the chore option, and added a test to ensure all of the options in any select questions in cz_conventional_commits has a keyboard key defined. Actually mocking out and asserting the shortcuts render in the terminal when the config option is enabled was a little too complicated IMO. I also updated the config tests to be aware of the new config option.

Lee-W commented on commitizen/cz/conventional_commits/conventional_commits.py:
```diff
@@ -45,55 +45,68 @@ def questions(self) -> List[Dict[str, Any]]:
                 "type": "list",
                 "name": "prefix",
                 "message": "Select the type of change you are committing",
+                "use_shortcuts": self.config.settings['use_shortcuts'],
```
How about make it `settings['cz_conventional_commits']['use_shortcuts']`? This configuration might not make sense in other cz rules.

zevisert replied:
Sure, conventional is the only pre-defined ruleset that uses a list select, but keep in mind that customized cz rules may also wish to use shortcuts.

Lee-W replied:
Yep, that sounds reasonable. Once the documentation is updated. I think we're good to merge it. Thanks for this awesome feature!

zevisert replied:
I moved this check from here to a filter in the actual commit, so that it automatically covers customized cz rules. Now use_shortcuts can apply to cz_customize in config files, and to classes defining `questions()` of their own.

Lee-W commented:
Thanks for your update! But might need your help to reformat the code through `./scripts/format`

zevisert commented:
Docs are updated, and I moved where we check the config for `use_shortcuts` to the `Commit` class, so that any `BaseCommitizen` class automatically gets access to this new feature. Thinking about my team and I, since we like to have `chore` included in our custom conventional commits rules.

zevisert commented:
Also, I mentioned it in the docs, but when there's no `key` provided, questionary conveniently falls back to numbering the options. This means that 3rd party templates aren't broken by this new setting! 🎉 

eg.; Using the [customize settings from the docs](https://commitizen-tools.github.io/commitizen/customization/#1-customize-in-configuration-file) with use_shortcuts enabled: 
```shell
 > poetry run cz commit
? Select the type of change you are committing (Use shortcuts or arrow keys)
 » 1) feature: A new feature.
   2) bug fix: A bug fix.
```

Lee-W commented:
didn't expect that handy feature. it's great. let's merge this one!
